dispatcher/helpers/send.py

```diff
@@ -0,0 +1,246 @@
+                "last_name": faker.last_name(),
+                "body_name": random.choice(BODIES_LIST),
+                "created_at": "yesterday or whatever",
+                "updated_at": "now more or less",
+                "aegee_experience": "I suck",
+                "ideal_su": "I get to suck",
+                "motivation": "I wanna suck",
+                "status": "totally snafu",
+                "email": "someone@example.org",
+                "gender": "Robot",
+                "date_of_birth": "1985-04-16",
+                "nationality": "Europe",
+                "meals": "Steak",
+                "allergies": "wallopers",
+                "number_of_events_visited": "over 9000",
+                "visa_required": "No",
+                "answers": [
+                    "ho-hoo, ho-hoo",
+                    "42",
+                    ],
+            },
+            "candidate": {
+                "first_name": faker.first_name(),
+                "last_name": faker.last_name(),
+            },
+            "position": {
+                "event_id": "42",
+                "name": "This gran C commissioner",
+            },
+            "body": """
+                <ul>
+                    <li><strong>Who is cool: </strong>Accountable people</li>
+                    <li><strong>Who is not: </strong>People hiding behind the excuse of 'volunteer'</li>
+                </ul>
+            """
+        }
+    }
+    return email
+
+
+if(RANDOM_AMOUNT_TEST):
+    amount = random.randrange(MIN_MSG,MAX_MSG)
+    for _ in range(amount):
+        email = generate_fake_payload()
+        channel.basic_publish(exchange='eml',
+                            routing_key='mail',
+                            body=json.dumps(email),
+                            properties=pika.BasicProperties(
+                                delivery_mode = pika.spec.PERSISTENT_DELIVERY_MODE
+                            ))
+        print(f" [x] Sent {email['subject']} (to {email['to']})")
+    print(f" Gee, I sent all  {amount} ")
+
+if(ALL_TEMPLATES_TEST):
+    templates_tested=0
+    for ms in MAIL_TEMPLATES:
+        for case in MAIL_TEMPLATES[ms]:
+            email = generate_fake_payload(MAIL_SUBJECTS[ms][case], MAIL_TEMPLATES[ms][case])
+            channel.basic_publish(exchange='eml',
+                                routing_key='mail',
+                                body=json.dumps(email),
+                                properties=pika.BasicProperties(
+                                    delivery_mode = pika.spec.PERSISTENT_DELIVERY_MODE
+                                ))
+            print(f" [x] Sent {email['subject']}")
+            templates_tested+=1
+    file_count = len([f for f in os.listdir("../templates/") if os.path.isfile(os.path.join("../templates/", f))])
+    if(file_count != templates_tested):
+        print()
+        print()
+        print(f"Tested {templates_tested} templates but the folder contains {file_count}")
```

Comment:
could we add a "os.Exit(1)" here?

I guess what I'm looking for, is a section in the README "how to dev locally?". And I'd say, if you modify the python, then you want to set ALL_TEMPLATES_TEST=false, and RANDOM_AMOUNT_TEST=true, but if you modify the templates, you surely want ALL_TEMPLATES_TEST=true.

When testing all templates, I think we should mention that we catch some errors (invalid jinja syntax, template not found) but that a lot of things are not covered by this test:
* validity of HTML generated: does it look legible? (also on mobile, and without images)
* validity of content: do we send the proper content: does a new candidature email contain the string "                            A new candidature to position"? I wonder if for this purpose (checking the content), some sort of golden tests would be nice: we don't have dynamic (dates) content, while we use random names for the fields, we could make generate_fake_payload static/deterministic.
* finally, that the service calling the dispatcher does pass enough data to the mailer, so that no field is empty.

Reply:
i still don't understand your points here, I read them once again but I am lost :) can you rephrase? also your bullet list to me seems unrelated to the first part of the comment

Reply:
fair.

>could we add a "os.Exit(1)" here?

right now, send.py always returns 0, regardless of whether tests passed or not. When we detect that one test case failed, we should return to the OS an exit code !=0 to notify it of a test failure. This can then be integrated with other tooling (in a fish terminal, this will show a red prompt for instance, to catch attention)


>I guess what I'm looking for, is a section in the README "how to dev locally?". ...

We should explain the usefulness of each option, and when one wants to use ALL_TEMPLATES_TEST and when one wants to use RANDOM_AMOUNT_TEST.



>When testing all templates,...

I would like to see written what this test can catch, and what this test doesn't cover at all. This way, a developer knows that they need to perform some extra manual testing to ensure that they didn't break anything: this script won't do it all for them (which is totally fine, as long as it's explicit)